src/algebraic_geometry/EllipticCurve/point.lean

```diff
@@ -0,0 +1,356 @@
+    line has a well-defined gradient $\ell = (y_1 - y_2) / (x_1 - x_2)$. The $x$-coordinate of
+    $P + Q$ is then the unique third solution of the equation obtained by substituting the line
+    $y = \ell(x - x_1) + y_1$ into the Weierstrass equation, and can be written down explicitly by
+    inspecting the $x^2$ terms of this equation as $x = \ell^2 + a_1\ell - a_2 - x_1 - x_2$. The
+    $y$-coordinate of $P + Q$, after applying the final negation that maps $y$ to $-y - a_1x - a_3$,
+    is precisely $y = -(\ell(x - x_1) + y_1) - x - a_3$.
+The group law on this set is then uniquely determined by these constructions. This file defines the
+group operations by explicit equations and proves that the set is an abelian group (TODO).
+
+## Main definitions
+
+ * `EllipticCurve.point`: a `K`-rational point on `E`.
+ * `EllipticCurve.point.add`: addition of two `K`-rational points on `E`.
+
+## Main statements
+
+ * `EllipticCurve.point.add_comm`: addition of two `K`-rational points on `E` is commutative.
+ * TODO: addition of two `K`-rational points on `E` is associative (HARD), and hence forms a group.
+
+## Notations
+
+ * `E⟮K⟯`: the group of `K`-rational points on an elliptic curve `E`.
+
+## Implementation notes
+
+The proofs of `EllipticCurve.point.weierstrass_dbl'`, `EllipticCurve.point.weierstrass_add'`, and
+`EllipticCurve.point.add_assoc` (TODO) are slow due to the sheer size of the polynomials involved,
+but they can be hastened considerably by breaking the computations into multiple steps if necessary.
+
+## References
+
+[J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, rational point, group law
+-/
+
+universes u v
+
+variables {F : Type u} [field F] (E : EllipticCurve F) {K : Type v} [field K] [algebra F K]
+
+namespace EllipticCurve
+
+/-- The proposition that an affine point $(x, y)$ lies in `E`. In other words, it satisfies the
+Weierstrass equation $y^2 + a_1xy + a_3y = x^3 + a_2x^2 + a_4x + a_6$. -/
+@[simp] def weierstrass (x y : K) : Prop :=
+y ^ 2 + ↑E.a₁ * x * y + ↑E.a₃ * y = x ^ 3 + ↑E.a₂ * x ^ 2 + ↑E.a₄ * x + ↑E.a₆
+
+variables (K)
+
+/-- The group of `K`-rational points `E⟮K⟯` on an elliptic curve `E` over `F`. This consists of the
+unique point at infinity `EllipticCurve.point.zero`and the affine points `EllipticCurve.point.some`
+satisfying the Weierstrass equation $y^2 + a_1xy + a_3y = x^3 + a_2x^2 + a_4x + a_6$. -/
+inductive point
+| zero
+| some (x y : K) (h : E.weierstrass x y)
+
+localized "notation E⟮K⟯ := point E K" in point
+
+variables {E K}
+
+namespace point
+
+instance : inhabited E⟮K⟯ := ⟨zero⟩
+
+/-!
+### Zero in `E⟮K⟯`
+
+Use `0` instead of `EllipticCurve.point.zero`.
+-/
+
+section zero
+
+instance : has_zero E⟮K⟯ := ⟨zero⟩
+
+@[simp] lemma zero_def : zero = (0 : E⟮K⟯) := rfl
+
+end zero
+
+/-!
+### Negation in `E⟮K⟯`
+
+Given `P : E⟮K⟯`, use `-P` instead of `neg P`.
```

Comment:
These warnings should appear in the docstring of the definitions (`neg` in this case) as well.